### PR TITLE
feat: fp32 vector to fp16/bf16 vector conversion for RESTful API

### PR DESCRIPTION
**Pull request by jiangyinzuo**
RESTful API. The influenced API are as follows:

- Handler. insert
- HandlerV1. insert/upsert
- HandlerV2. insert/upsert/search

We do not modify search API in Handler/HandlerV1 because they do not support fp16/bf16 vectors.

module github.com/milvus-io/milvus/pkg:

Add `Float32ArrayToBFloat16Bytes()`, `Float32ArrayToFloat16Bytes()` and
`Float32ArrayToBytes()`. These method will be used in GoSDK in the
future.

issue: #37448

### CODE REVIEW

File: internal/distributed/proxy/httpserver/wrap_request.go [hunk 1]
```diff
@@ -320,18 +367,6 @@ func vector2Bytes(vectors [][]float32) []byte {
...
-func serializeVectors(fv []float32) []byte {
-	data := make([]byte, 0, 4*len(fv)) // float32 occupies 4 bytes
-	buf := make([]byte, 4)
-	for _, f := range fv {
-		binary.LittleEndian.PutUint32(buf, math.Float32bits(f))
-		data = append(data, buf...)
-	}
-	return data
-}
-
```
Comment (jiangyinzuo) [inline internal/distributed/proxy/httpserver/wrap_request.go:334]: duplicate function with `serialize`, so I removed it.

File: internal/distributed/proxy/httpserver/utils.go [hunk 2]
```diff
@@ -809,9 +805,25 @@ func anyToColumns(rows []map[string]interface{}, validDataMap map[string][]bool,
...
 			case schemapb.DataType_BFloat16Vector:
-				nameColumns[field.Name] = append(nameColumns[field.Name].([][]byte), candi.v.Interface().([]byte))
+				switch candi.v.Interface().(type) {
+				case []byte:
+					nameColumns[field.Name] = append(nameColumns[field.Name].([][]byte), candi.v.Interface().([]byte))
+				case []float32:
+					vec := typeutil.Float32ArrayToBFloat16Bytes(candi.v.Interface().([]float32))
+					nameColumns[field.Name] = append(nameColumns[field.Name].([][]byte), vec)
+				default:
+					return nil, fmt.Errorf("invalid type(%v) of field(%v) ", field.DataType, field.Name)
```
Comment (smellthemoon) [inline internal/distributed/proxy/httpserver/utils.go:825]: same here, please use merr to return the error.

File: internal/distributed/proxy/httpserver/utils.go [hunk 3]
```diff
@@ -1091,6 +1092,26 @@ func serializeByteVectors(vectorStr string, dataType schemapb.DataType, dimensio
...
 
+func serializeFp16OrBf16Vectors(vectorStr string, dataType schemapb.DataType, dimension int64, serializeFunc func([]float32) []byte) ([][]byte, error) {
+	// try to unmarshal as [][]float32 first to make sure `[[3, 3]]` is [][]float32 instead of [][]byte
+	fp32Values := make([][]float32, 0)
+	err := json.Unmarshal([]byte(vectorStr), &fp32Values)
+	if err == nil {
+		values := make([][]byte, 0)
+		for _, vectorArray := range fp32Values {
+			if int64(len(vectorArray)) != dimension {
+				vecStr, _ := json.MarshalToString(vectorArray)
```
Comment (smellthemoon) [inline internal/distributed/proxy/httpserver/utils.go:1103]: the vecStr seems only used to report err? It is recommended not to use marshal one more time to get vecStr, but to pass in other valid error information, such as field name, to prompt the user to pass in the correct data.

**Review by zhuwenxing — APPROVED** [review]

**Comment by jiangyinzuo** [comment]
rerun go-sdk

File: internal/distributed/proxy/httpserver/utils.go [hunk 4]
```diff
@@ -367,27 +365,25 @@ func checkAndSetData(body string, collSchema *schemapb.CollectionSchema) (error,
...
 					}
 					vectorStr := gjson.Get(data.Raw, fieldName).Raw
-					var vectorArray []byte
-					err := json.Unmarshal([]byte(vectorStr), &vectorArray)
-					if err != nil {
-						return merr.WrapErrParameterInvalid(schemapb.DataType_name[int32(fieldType)], dataString, err.Error()), reallyDataArray, validDataMap
+					// unmarshal []float32 first to make sure `[3, 3]` is []float instead of []byte
+					var float32Array []float32
+					err := json.Unmarshal([]byte(vectorStr), &float32Array)
+					if err == nil {
```
Comment (czs007) [inline internal/distributed/proxy/httpserver/utils.go:377]: Is there a chance of a false positive occurring here?

File: internal/distributed/proxy/httpserver/utils.go [hunk 5]
```diff
@@ -1118,6 +1119,25 @@ func serializeByteVectors(vectorStr string, dataType schemapb.DataType, dimensio
 	return values, nil
 }
 
+func serializeFp16OrBf16Vectors(vectorStr string, dataType schemapb.DataType, dimension int64, serializeFunc func([]float32) []byte) ([][]byte, error) {
+	// try to unmarshal as [][]float32 first to make sure `[[3, 3]]` is [][]float32 instead of [][]byte
+	fp32Values := make([][]float32, 0)
+	err := json.Unmarshal([]byte(vectorStr), &fp32Values)
+	if err == nil {
```
Comment (czs007) [inline internal/distributed/proxy/httpserver/utils.go:1126]: Is there a chance of a false positive occurring here?
the performance here is likely to deteriorate, isn't it?

**Comment by jiangyinzuo** [comment]
```bash
go test -bench=BenchmarkSerialize -run=^$
```

```
goos: linux
goarch: amd64
pkg: github.com/milvus-io/milvus/internal/distributed/proxy/httpserver
cpu: Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz
BenchmarkSerialize_FloatVectors_Baseline-40                            7         144641969 ns/op
BenchmarkSerialize_FloatVectors-40                                    10         108097462 ns/op
BenchmarkSerialize_FloatVectors_Float16-40                             7         148711147 ns/op
BenchmarkSerialize_FloatOrByteVectors_Fp32-40                          7         158311570 ns/op
BenchmarkSerialize_FloatOrByteVectors_Byte-40                         58          19690757 ns/op
BenchmarkSerialize_FloatOrByteVectors_Fp32_UnmashalTwice-40            7         151955889 ns/op
BenchmarkSerialize_FloatOrByteVectors_Byte_UnmashalTwice-40           37          33288434 ns/op
BenchmarkSerialize_ByteVectors-40                                     55          21864953 ns/op
PASS
ok      github.com/milvus-io/milvus/internal/distributed/proxy/httpserver       17.815s
```

**Comment by jiangyinzuo** [comment]
rerun cpp-unit-test

**Comment by jiangyinzuo** [comment]
rerun go-sdk

**Pull request merged**